firedrake/assemble.py

```diff
@@ -161,14 +161,24 @@ def _make_vector(test):
 
 def _assemble_expr(expr, tensor, *, bcs, diagonal, assemble_now, **kwargs):
     bcs = _preprocess_bcs(bcs)
-    # TODO: Reimplement this when PyOP2 functionality added.
-    # if "parloops" in expr._cache:
-    #     parloops = expr._cache["parloops"]
-    # else:
-    #     parloops = _make_parloops(expr, tensor, bcs=bcs, diagonal=diagonal, **kwargs)
-    #     expr._cache["parloops"] = parloops
-    # parloop.compute(out=tensor.dat)
-    parloops = _make_parloops(expr, tensor, bcs=bcs, diagonal=diagonal, **kwargs)
+
+    # We cache the parloops on the form but keep track of the tensor. If the
+    # tensor is changed then we need to regenerate the parloop (until parloops
+    # stop depending on data).
+    # When we generate a new parloop the old one is removed from the cache to
+    # prevent leaking memory.
+    cache_key = "parloops"
+    if cache_key in expr._cache:
+        cached_tensor, parloops = expr._cache[cache_key]
+        if cached_tensor is not tensor:
+            parloops = None
+            del expr._cache[cache_key]
+    else:
+        parloops = None
+
+    if not parloops:
+        parloops = _make_parloops(expr, tensor, bcs=bcs, diagonal=diagonal, **kwargs)
+        expr._cache[cache_key] = (tensor, parloops)
```

Comment:
I guess this is a stop-gap, but it might be surprising that a form holds on to its output tensor. Would it work to hold a `weakref` here?

Reply:
I guess. The tensor's data is still tied to the parloop anyway so not sure it really matters.

Reply:
Oh yeah, in which case, don't bother.